dataservices/dataservices.go

```diff
@@ -0,0 +1,239 @@
+)
+
+const (
+	missingPermissionsError = "missing required permissions"
+	missingDataError        = "missing data to process"
+	gettingDataError        = "there was an error getting your data"
+
+	dataservicesName = "dataservices"
+	useridParamName  = "userid"
+)
+
+var (
+	log = logger.Log.GetNamed(dataservicesName)
+)
+
+func main() {
+	//TODO: from config
+	log.Fatal(NewDataServiceClient().Run(":8077"))
+}
+
+//DataServiceClient for the data service
+type DataServiceClient struct {
+	api              *rest.Api
+	dataStore        store.Store
+	validateToken    user.ChainedMiddleware
+	attachPermissons user.ChainedMiddleware
+	resolveGroupID   user.ChainedMiddleware
+}
+
+//NewDataServiceClient returns an initialised client
+func NewDataServiceClient() *DataServiceClient {
+	log.Info(version.Long())
+
+	userClient := user.NewServicesClient()
+	userClient.Start()
+
+	return &DataServiceClient{
+		api: rest.NewApi(),
+		//TODO: from config
+		dataStore:        store.NewMongoStore("deviceData"),
+		validateToken:    user.NewAuthorizationMiddleware(userClient).ValidateToken,
+		attachPermissons: user.NewMetadataMiddleware(userClient).GetPermissons,
+		resolveGroupID:   user.NewMetadataMiddleware(userClient).GetGroupID,
+	}
+
+}
+
+//Run will run the service
+func (client *DataServiceClient) Run(URL string) error {
+	client.api.Use(rest.DefaultDevStack...)
+	client.api.Use(&rest.GzipMiddleware{})
+
+	router, err := rest.MakeRouter(
+		rest.Get("/version", client.GetVersion),
+		rest.Get("/data/:userid/:datumid", client.validateToken(client.resolveGroupID((client.GetData)))),
+		rest.Post("/dataset/:userid", client.validateToken(client.attachPermissons(client.resolveGroupID(client.PostDataset)))),
+		rest.Get("/dataset/:userid", client.validateToken(client.attachPermissons(client.resolveGroupID(client.GetDataset)))),
+		rest.Post("/blob/:userid", client.validateToken(client.attachPermissons(client.resolveGroupID(client.PostBlob)))),
+	)
+	if err != nil {
+		log.Fatal(err)
+	}
+	client.api.SetApp(router)
+
+	return http.ListenAndServe(URL, client.api.MakeHandler())
+}
+
+//checkPermisson will check that we have the expected permisson
+func checkPermisson(r *rest.Request, expected user.Permission) bool {
+	//TODO: fill in the details
+	if permissions := r.Env[user.PERMISSIONS].(*user.UsersPermissions); permissions != nil {
+		return true
+	}
+	return false
+}
+
+//GetVersion will return the current API version
+func (client *DataServiceClient) GetVersion(w rest.ResponseWriter, r *rest.Request) {
+	w.WriteJson(version.Long())
+	return
+}
+
+//PostDataset will process a posted dataset for the requested user if permissons are sufficient
+func (client *DataServiceClient) PostDataset(w rest.ResponseWriter, r *rest.Request) {
+
+	userid := r.PathParam(useridParamName)
+
+	log.AddTrace(userid)
+
+	if checkPermisson(r, user.Permission{}) {
+
+		groupID := r.Env[user.GROUPID]
+
+		if r.ContentLength == 0 || groupID == "" {
```

Comment:
I'd separate these two out.  A missing groupID is not missing data.  It is an internal error (since it should have failed earlier in the middleware chain.